validator/client/src/main/java/tech/pegasys/teku/validator/client/BeaconProposerPreparer.java

```diff
@@ -15,63 +15,107 @@
-import tech.pegasys.teku.validator.client.loader.OwnedValidators;
+import tech.pegasys.teku.validator.client.ProposerConfig.Config;
+import tech.pegasys.teku.validator.client.proposerconfig.ProposerConfigProvider;
 
 public class BeaconProposerPreparer implements ValidatorTimingChannel {
+  private static final Logger LOG = LogManager.getLogger();
+
   private final ValidatorApiChannel validatorApiChannel;
   private final ValidatorIndexProvider validatorIndexProvider;
-  private final OwnedValidators validators;
+  private final ProposerConfigProvider proposerConfigProvider;
   private final Spec spec;
-  private final Optional<? extends Bytes20> feeRecipient;
+  private final Optional<? extends Bytes20> defaultFeeRecipient;
   private boolean firstCallDone = false;
 
   public BeaconProposerPreparer(
       ValidatorApiChannel validatorApiChannel,
       ValidatorIndexProvider validatorIndexProvider,
-      Optional<? extends Bytes20> feeRecipient,
-      OwnedValidators validators,
+      ProposerConfigProvider proposerConfigProvider,
+      Optional<? extends Bytes20> defaultFeeRecipient,
       Spec spec) {
     this.validatorApiChannel = validatorApiChannel;
     this.validatorIndexProvider = validatorIndexProvider;
-    this.validators = validators;
-    this.feeRecipient = feeRecipient;
+    this.proposerConfigProvider = proposerConfigProvider;
+    this.defaultFeeRecipient = defaultFeeRecipient;
     this.spec = spec;
   }
 
   @Override
   public void onSlot(UInt64 slot) {
     if (slot.mod(spec.getSlotsPerEpoch(slot)).isZero() || !firstCallDone) {
       firstCallDone = true;
+
+      SafeFuture<Optional<ProposerConfig>> proposerConfigFuture =
+          proposerConfigProvider.getProposerConfig();
+
       validatorIndexProvider
-          .getValidatorIndices(validators.getPublicKeys())
+          .getValidatorIndexesByPublicKey()
           .thenApply(
-              integers ->
-                  integers.stream()
-                      .map(
-                          index ->
-                              new BeaconPreparableProposer(
-                                  UInt64.valueOf(index), getFeeRecipient()))
-                      .collect(Collectors.toList()))
+              blsPublicKeyIntegerMap ->
+                  proposerConfigFuture
+                      .thenApply(
+                          proposerConfig ->
+                              buildBeaconPreparableProposerList(
+                                  proposerConfig, blsPublicKeyIntegerMap))
+                      .exceptionally(
+                          throwable -> {
+                            LOG.warn(
+                                "An error occurred while obtaining proposer config", throwable);
+                            return buildBeaconPreparableProposerList(
+                                Optional.empty(), blsPublicKeyIntegerMap);
+                          })
+                      .join())
```

Comment:
we shouldn't need a join here.  I suspect we want the first `thenApply` to actually just be a `thenCompose` which would let this return the future rather than calling join.

`thenApply` is like the future version of `Optional.map` whereas `thenCompose` is like `Optional.flatMap` if that helps the mental model at all (it may not). :)

Reply:
ah yes, most of the time I get it right but this time I fought a bit more. I was sure there must have been a clean version of it :)

Reply:
thanks for the mental model hint. Definitely helps as a shortcut without the need to replay the entire story in your head :smile: